### PR TITLE
[Cleanup] Replace ttnn.gelu_bw with ttnn.experimental.gelu_bw

### LINKED ISSUE #54152: [ttnn] Duplicate gelu_bw ops, and `approximate` is a string instead of GeluVariant
## 1. There are two `gelu_bw` ops

`ttnn::gelu_bw` and `ttnn::experimental::gelu_bw` both implement the GELU backward pass, each with its own device operation, program factory and compute kernels.

|  | `ttnn.gelu_bw` | `ttnn.experimental.gelu_bw` |
|---|---|---|
| Declaration | `ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.hpp:428` | `ttnn/cpp/ttnn/operations/experimental/unary_backward/gelu_backward/gelu_backward.hpp:12` |
| Device op | `eltwise/unary_backward/gelu_bw/device/` (`GeluBwDeviceOperation`) | `experimental/unary_backward/gelu_backward/device/` (`GeluBackwardDeviceOperation`, via `ttnn::prim::gelu_bw`) |
| Returns | `std::vector<std::optional<Tensor>>` (1 element; a list in Python) | `Tensor` |
| `approximate` validation | `TT_FATAL` unless `"none"`/`"tanh"` (`unary_backward.cpp:1569`), then collapsed to a `bool` | none — any string other than `"tanh"` silently selects the polynomial kernel (`gelu_backward_program_factory.cpp:179`), and the raw string is part of the program-cache key |
| Program factory | legacy `ProgramDescriptor` | Metal 2.0 `ProgramSpec` / `DataflowBuffer` |
| Kernels | `eltwise_bw_gelu_poly.cpp`, `eltwise_bw_gelu_tanh.cpp`, `eltwise_bw_gelu_tanh_fp32.cpp` | `eltwise_bw_gelu_poly.cpp`, `eltwise_bw_gelu_approx_tanh.cpp` (no dedicated fp32 tanh kernel) |
| Rounding | `BF16_ROUNDING_MODE` → round-nearest-even for BFLOAT16 output | none |
| dtype validation | explicit whitelist (BFLOAT16 / FLOAT32 / BFLOAT8_B / BFLOAT4_B) on grad and input | only `grad.dtype() == input.dtype()` |
| `memory_config` | ignored — `empty_like(grad)` at `unary_backward.cpp:1563-1568` runs before the ternary, so `output_mem_config` is unreachable | honored |
| Tests | `tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_gelu.py`, `tests/ttnn/unit_tests/operations/eltwise/test_gelu_bw_main_ulp.py`, `tests/ttnn/unit_tests/gtests/test_gelu_bw_main_ulp.cpp` | `.../backward/test_backward_gelu_fused.py`, `.../test_gelu_bw_ulp.py`, `gtests/test_gelu_bw_ulp.cpp` |

The two `eltwise_bw_gelu_poly.cpp` files are near-verbatim copies — they differ only by the copyright header and the circular-buffer → `DataflowBuffer` API migration. Fixes and accuracy work have to be applied twice, or one copy silently drifts.

## 2. `approximate` is a string, not a `GeluVariant`

The forward op takes an enum:

```cpp
// ttnn/cpp/ttnn/operations/eltwise/unary/unary.hpp:19, :131
enum class GeluVariant { ACCURATE, FAST_LUT, TANH };

Tensor gelu(const Tensor&, operations::unary::GeluVariant variant = GeluVariant::ACCURATE, ...);
```

Both backward ops take a string:

```cpp
std::vector<std::optional<Tensor>> gelu_bw(const Tensor&, const Tensor&, const std::string& approximate, ...);
Tensor ttnn::experimental::gelu_bw(const Tensor&, const Tensor&, const std::string& approximate = "none", ...);
```

So any caller holding a `GeluVariant` has to convert it back to a string to build the gradient. tt-train currently carries a helper for exactly this (`gelu_bw_approximate()` in `tt-train/sources/ttml/ops/unary_ops.cpp`), and it is silent rather than diagnostic on the experimental path because that op does not validate the string.

(`ttnn::bias_gelu_bw` has the same string parameter.)

## Ask

1. Pick the surviving `gelu_bw` and remove or alias the other. Neither is a superset today: the experimental one has the Metal 2.0 program factory, the non-experimental one has the fp32 tanh kernel, RNE rounding, the dtype whitelist and `approximate` validation. Whichever survives needs all of it.
2. Change `approximate` to `GeluVariant` for parity with `ttnn::gelu`.

Note: There is a mismatch with `FAST_LUT` variant being forward-only. In practice, LUT variant is quite inaccurate, and it would not be suitable for training workload (probably best to just reject it in gelu_bw). 

Related: #48710 (closed).

### PR DESCRIPTION
**Pull request by nmauriceTT**
### PR Category
Cleanup

### Summary
Closes #54152

- Make `ttnn.gelu_bw` use the feature-complete implementation previously under `ttnn.experimental.gelu_bw`, then remove the duplicate experimental op.
- Return value is kept consistent with `ttnn.gelu_bw`: vector of optional tensors. 
- Consolidate GELU backward coverage around exhaustive allclose checks and special values.
- Retain C++ and Python ULP regression suites for both approximations.
- `ttnn.gelu_bw` now takes `ttnn.GeluVariant` as input argument (consistent with `ttnn.gelu`)

### Notes for reviewers
- The old experimental and new public tanh kernels use the same GELU derivative. For BF16, the replacement explicitly applies nearest-even destination rounding to intermediate add/sub operations; the previous experimental kernel used the default rounding mode.
- Unlike what was stated in the nanobind docstring, neither `ttnn.gelu_bw` nor `ttnn.experimental.gelu_bw` support row-major inputs. 

**Note**: This is a breaking change for `ttnn.gelu_bw` as it changes its API. I don't think this is used in lots of places. If backward compatibility is desired, I can add a backward-compatible binding.

- Sanity fails on unrelated tests (DeepSeek tests seem to be failing on a infra-related problem). 
- L2 Nightly: tt-train tests fail on grpo tests (unrelated and also happen on main)

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nmaurice/unify-gelu-approximations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:nmaurice/unify-gelu-approximations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nmaurice/unify-gelu-approximations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:nmaurice/unify-gelu-approximations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nmaurice/unify-gelu-approximations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nmaurice/unify-gelu-approximations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nmaurice/unify-gelu-approximations)
- Validation rerun on 2026-09-08 after the build fix: the TTNN target builds successfully. The full GELU backward suite passes (27 passed, 5 documented xfailed). Exhaustive BF16 eltwise tester summaries with grad=1: Accurate max/mean/useful-average ULP = 3.49669/0.0160419/0.0722417; Tanh = 349.197/0.539921/2.45719. Temporary tester entries were restored.

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/gelu_bw/device/gelu_bw_device_operation_types.hpp [hunk 1]
```diff
@@ -12,7 +12,7 @@ namespace ttnn::operations::unary_backward::gelu_bw {
 struct GeluBwParams {
     const tt::tt_metal::DataType output_dtype = tt::tt_metal::DataType::INVALID;
     const tt::tt_metal::MemoryConfig output_memory_config;
-    const bool approximate = false;
+    const std::string approximate = "none";
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/eltwise/unary_backward/gelu_bw/device/gelu_bw_device_operation_types.hpp]: Updated argument type to `ttnn.GeluVariant`.

File: tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_gelu.py [hunk 2]
```diff
@@ -1,278 +1,232 @@
...
-        "(no new entries). A new entry means the buffers/mode were wrongly folded into the hash."
-    )
-
-    logger.debug("gelu_bw program-cache regression: 2 entries, both modes correct across cache hits")
 
-    device.disable_and_clear_program_cache()
+    def fresh_inputs(seed):
+        torch.manual_seed(seed)
+        input_data = (torch.rand(shape, dtype=torch.bfloat16) * 10 - 5).detach()
+        grad_data = torch.ones(shape, dtype=torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_gelu.py]: Added cache related tests with various configurations.

File: tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_gelu.py [hunk 3]
```diff
@@ -1,278 +1,232 @@
...
-    tt_output_tensor_on_device = [input_grad]
+def _bf16_tolerance(approximate):
+    return (4.9e-2, 1.5e-2) if approximate == "tanh" else (2e-2, 9e-3)
+
 
-    golden_function = ttnn.get_golden_function(ttnn.gelu_bw)
-    golden_tensor = golden_function(grad_data, in_data, approximate=approximate)
-    tt_out = ttnn.to_torch(tt_output_tensor_on_device[0])
+@pytest.mark.parametrize("approximate,torch_dtype,ttnn_dtype,rtol,atol", EXHAUSTIVE_TEST_CASES)
+def test_gelu_bw_exhaustive_allclose(device, approximate, torch_dtype, ttnn_dtype, rtol, atol):
```
[the remark this answers is not in the document]
Comment (nmauriceTT) [reply tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_gelu.py]: Added tests with unaligned shapes.

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_nanobind.cpp [hunk 4]
```diff
@@ -452,7 +452,7 @@ void bind_unary_backward_float_with_default(
     const std::string& note = "") {
     auto doc = fmt::format(
         R"doc(
-        {5}
+        {2}
```
Comment (jmalone-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_nanobind.cpp]: If I'm understanding, this changes from `description, supported_dtype, note` used in the docs to `parameter_name_a, parameter_a_doc, parameter_a_value`.
Seems like this might be unintentional - can you double check?
(same comment about similar change below)

File: tests/ttnn/profiling/ops_for_profiling.py [hunk 5]
```diff
@@ -154,19 +154,19 @@ def clamp_max_bw(x, y):
...
 
 
 def gelu_bw_tanh(x, y):
-    ttnn.gelu_bw(x, y, approximate="tanh")
+    ttnn.gelu_bw(x, y, variant=ttnn.GeluVariant.Tanh)
 
 
 def bias_gelu_unary_bw_none(x, y):
-    ttnn.bias_gelu_bw(x, y, bias=3.1, approximate="none")
+    ttnn.bias_gelu_bw(x, y, bias=3.1, variant=ttnn.GeluVariant.Accurate)
```
Comment (jmalone-tt) [inline tests/ttnn/profiling/ops_for_profiling.py:165]: I think you might be missing updates for bias_gelu_bw in this PR (still binds `approximate` instead of `variant`)
Won't break in CI, but I expect it would break if run manually

**Review by jmalone-tt — COMMENTED** [review]
Looks good overall, but a couple things to fix before merging

**Review by vmelnykovTT — APPROVED** [review]
LGTM!

**Review by philei-tt — APPROVED** [review]

File: tests/ttnn/profiling/ops_for_profiling.py (hunk 5 above)
Comment (nmauriceTT) [reply tests/ttnn/profiling/ops_for_profiling.py:165]: Thanks. I have updated `ttnn.bias_gelu_bw` calls as well. (including adding backward compatibility with string 'approximate').

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_nanobind.cpp (hunk 4 above)
Comment (nmauriceTT) [reply ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_nanobind.cpp]: Good catch. I have updated the docstring + built & visualized matching docs.

**Review by bbradelTT — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]

**Review by dgomezTT — APPROVED** [review]
